compiler/src/iree/compiler/Codegen/Utils/Utils.h

```diff
@@ -225,6 +225,15 @@ void sinkOpsInCFG(const SmallVector<Operation *> &allocs,
 // the inputs.
 bool hasFusedLeadingOp(linalg::LinalgOp rootOp);
 
+/// Replace the memory space of all memref types recursively in the given
+/// `root` operation. The replace function takes the current memory space
+/// attribute and returns the new memory space attribute. If no replacement is
+/// returned, this type does not need to be replaced. A null attribute
+/// replacement will remove the memory space attribute.
+LogicalResult replaceMemRefMemorySpace(
```

Comment:
This is a very specific pattern. Does it need to be here.

Also broad level... why does the memory space munging need to happen. I am not sure type munging this way is easy to maintain and develop. Broadly, type is mostly for verification. It is not really the intent to use it to carry analysis state.